### PR TITLE
fix Issue 23387 - ImportC: identical structs defined in two C files l…

### PR DESCRIPTION
**Pull request by WalterBright**
…ead to duplicate .init symbol on macOS

Note that the test should should compile them separately:
```
dmd imports/maker.i -c
dmd imports/freer.i -c
dmd test23387.d -c
dmd test23387.o freer.o maker.o
```
But I don't see a way to do that with the test runner.

### CODE REVIEW

File: compiler/test/runnable/test23387.d [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// https://issues.dlang.org/show_bug.cgi?id=23387
+
+/+ maker.i
+typedef struct Foo *FooRef;
+struct Foo {
+    int x;
+};
+FooRef make_foo(void);
++/
+import imports/maker;
```
Comment (thewilsonator) [inline compiler/test/runnable/test23387.d:13]: `import imports_._maker;`

File: compiler/test/runnable/test23387.d [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+
+/+ freer.i
+typedef struct Foo *FooRef;
+struct Foo {
+    int x;
+};
+void free_foo(FooRef foo);
++/
+import imports/freer;
```
Comment (thewilsonator) [inline compiler/test/runnable/test23387.d:23]: ditto

File: compiler/test/runnable/test23387.d (hunk 1 above)
Comment (WalterBright) [reply compiler/test/runnable/test23387.d:13]: what does that do?

Comment (thewilsonator) [reply compiler/test/runnable/test23387.d:13]: you have a slash instead of a dot

Comment (WalterBright) [reply compiler/test/runnable/test23387.d]: That's what I get for switching back and forth between C and D!

**Comment by thewilsonator** [comment]
This is a backend change, cc @ibuclaw @kinke.

**Review by dkorpel — DISMISSED** [review]

File: compiler/src/dmd/tocsym.d [hunk 3]
```diff
@@ -655,7 +655,24 @@ Symbol *toInitializer(AggregateDeclaration ad)
...
         {
             auto stag = fake_classsym(Id.ClassInfo);
-            auto s = toSymbolX(ad, "__init", SC.extern_, stag.Stype, "Z");
+
+            Symbol* s;
+
+            Module m = ad.getModule();
+            if (m.filetype == FileType.c)
+            {
+                /* For ImportC structs, the module names are stripped from thje mangled name.
```
Comment (ibuclaw) [inline compiler/src/dmd/tocsym.d:664]:
```suggestion
                /* For ImportC structs, the module names are stripped from the mangled name.
```

**Comment by ibuclaw** [comment]
Why is dmd even generating this symbol for C sources?  The initializer is all zeros, no need to store that as a run-time symbol.

**Comment by RazvanN7** [comment]
ping @WalterBright

File: compiler/test/runnable/test23387.d [hunk 4]
```diff
@@ -0,0 +1,30 @@
+/* COMPILED_SEPARATELY:
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/test/runnable/test23387.d]: fixed

File: compiler/src/dmd/tocsym.d (hunk 3 above)
Comment (WalterBright) [reply compiler/src/dmd/tocsym.d]: fixed

**Comment by WalterBright** [comment]
I don't know what the test runner is asking for:
```
Test 'runnable/test23387.d' failed. The logged output:
/tmp/cirrus-ci-build/generated/freebsd/release/64/dmd -conf= -m64 -Irunnable  -fPIC  -odtest_results/runnable -c  runnable/test23387.d
/tmp/cirrus-ci-build/generated/freebsd/release/64/dmd -conf= -m64 -Irunnable  -fPIC  -odtest_results/runnable -c  runnable/imports/maker.i
/tmp/cirrus-ci-build/generated/freebsd/release/64/dmd -conf= -m64 -Irunnable  -fPIC  -odtest_results/runnable -c  runnable/imports/freer.i
/tmp/cirrus-ci-build/generated/freebsd/release/64/dmd -conf= -m64  -fPIC  -odtest_results/runnable -oftest_results/runnable/test23387_0 test_results/runnable/test23387.o test_results/runnable/maker.i test_results/runnable/freer.i
Error: cannot find input file `test_results/runnable/maker.i`
import path[0] = /tmp/cirrus-ci-build/compiler/test/../../druntime/import
import path[1] = /tmp/cirrus-ci-build/compiler/test/../../../phobos
==============================
Test 'runnable/test23387.d' failed: Expected rc == 0, but exited with rc == 1
```

**Comment by WalterBright** [comment]
@dkorpel do you know what is going on with the test runner?

**Comment by adamdruppe** [comment]
The C symbols shouldn't be separate anyway either. This is another case where the root fix would solve several bugs at once, but by adding more and more special cases for individual issues, the tech debt grows instead of shrinks as issues are closed.

**Comment by WalterBright** [comment]
@adamdruppe I understand what you mean, but do not believe it is necessary.

**Comment by dkorpel** [comment]
> @dkorpel do you know what is going on with the test runner?

No, but I can take a closer look tomorrow

**Comment by dkorpel** [comment]
The test runner doesn't replace the separately compiled ".i" extension to ".o", it only replaces ".d". I added a commit that should fix it.

**Review by dkorpel — APPROVED** [review]

**Pull request merged**